### ISSUE TITLE
Security policies not working if Pods are deployed before KubeArmor

### ISSUE THREAD

**Issue by oneiro-naut** [kubearmor/KubeArmor#163]
Security policies can't be deleted, modified and added to the pods which were deployed prior to KubeArmor deployment. #162 is a result of this issue. The following steps were ran on the kubearmor-dev vagrant VM.
Steps to reproduce the bug:
1. Deploy multiubuntu pods.
`vagrant@kubearmor-dev:~$ kubectl apply -f KubeArmor/examples/multiubuntu/`
2. Deploy KubeArmor.
`vagrant@kubearmor-dev:~$ kubectl apply -f KubeArmor/deployments/docker/`
3. Apply any security policy to multiubuntu pods.
```
vagrant@kubearmor-dev:~$ kubectl apply -f KubeArmor/examples/multiubuntu/security-policies/ksp-group-1-proc-path-block.yaml 
kubearmorpolicy.security.accuknox.com/ksp-group-1-proc-path-block created
```
4. Check if policy is enforced. Its not being enforced (we don't get any permission denied in this example).
```
vagrant@kubearmor-dev:~$ kubectl exec -ti -n multiubuntu ubuntu-1-deployment-5c475fb4fb-sq6xv -- /bin/bash
root@ubuntu-1-deployment-5c475fb4fb-sq6xv:/# /bin/sleep 1
```
NOTE: 
- kubectl shows deleted and created events but the actual AppArmor profiles in `/etc/apparmor.d/` aren't being modified or enforced. 
- The same happens if KubeArmor **restarts** since these are essentially the same conditions, ie pods being up before KubeArmor itself is up.

**Comment by geyslan** [comment]
I have noticed these behaviors. Thanks for creating the issue.

**Comment by oneiro-naut** [comment]
The problem seems to be happening because of this line in particular:
https://github.com/accuknox/KubeArmor/blob/a584a021a96d4d7c9c994f01ee15f9af1ec214da/KubeArmor/core/dockerHandler.go#L118
In the case of restart this value remains an empty string. For some reason the response received by the DockerClient doesn't contain anything for field AppArmorProfile. And this field is later is used to initialize other security policy related Data structures inside KubeArmor. This happens only after restarts so it might be related to Docker API's working.

**Comment by oneiro-naut** [comment]
Hi @nam-jaehyun. For the pods deployed before KubeArmor was deployed we don't get their "added" container event from the dockerHandler. The reason might be that the event is generated earlier than KubeArmor start and dockerClient is unable to fetch it. I have come up with an idea and would like to know your thoughts on this. The idea is that when we initialize the KubeArmor daemon we add 3 extra steps:

1. We fetch the list of current containers using dockerClient's ContainerList function. 
2. We filter out already running containers by comparing the "created" timestamp from the fetched list of containers with KubeArmor's own created timestamp.
3. We loop through this list and for each container in list we fetch containerInfo using ContainerInspect here[1] and add to KubeArmor daemon's container list[2].
[1]https://github.com/accuknox/KubeArmor/blob/1864d6ae9debea75ee2c07b22757e818f2fcbea4/KubeArmor/core/dockerHandler.go#L93
[2]https://github.com/accuknox/KubeArmor/blob/1864d6ae9debea75ee2c07b22757e818f2fcbea4/KubeArmor/core/kubeArmor.go#L49   
As of now we perform ContainerInspect when we get "start", "stop" and "destroy" events from Docker. But if containers start before KubeArmor starts the "start" events are all lost and never reported. So the idea is that they are still in the list of running containers for once(ie initialization) we can just fetch them from the list of containers.

**Comment by nam-jaehyun** [comment]
@oneiro-naut The key problem is that we cannot figure out the PID and MNT namespaces for already-deployed containers.
In other words, even though we know the container information deployed earlier than KubeArmor, we cannot map the system events with the discovered container information.

**Comment by oneiro-naut** [comment]
@nam-jaehyun Does this mean that even if we are able to enforce the policies on these pods we wont be able to monitor them? So we will need to make changes in the KubeArmor monitor as the final step?

**Comment by nam-jaehyun** [comment]
@oneiro-naut Yes. So, I created a new issue related to the monitor [#171](https://github.com/accuknox/KubeArmor/issues/171)

**Comment by oneiro-naut** [comment]
> @oneiro-naut The key problem is that we cannot figure out the PID and MNT namespaces for already-deployed containers.
> In other words, even though we know the container information deployed earlier than KubeArmor, we cannot map the system events with the discovered container information.

Are my steps to get the container information correct. Because we would still need them to enforce the policies on these already-deployed containers?

**Comment by nam-jaehyun** [comment]
@oneiro-naut In the current implementation, KubeArmor maintains the 'containers' list using Containerd and Docker APIs and it has all containers (even already deployed containers). At the same time, KubeArmor maintains all pods using K8s APIs and updates the namespace and pod information of containers in the 'containers' list. Thus, we already have all information related to containers. The only missing part is the mapping of PID/mnt namespaces and container IDs.

**Comment by nam-jaehyun** [comment]
Need to find a way to map the already deployed container IDs with the actual PID and MNT namespaces of the containers.
I think we need to update the system monitor (in particular, pid_ns_map management) to address this issue.

**Comment by oneiro-naut** [comment]
Hi @nam-jaehyun . The reason for not getting already deployed containers seems to be caused because we are missing execve event for pid == 1 (ie entrypoint execv for containers) which happens earlier and is therefore missed. https://github.com/kubearmor/KubeArmor/blob/b6df53ce1a910224c3aeba117361e6b9ecf52f14/KubeArmor/BPF/system_monitor.c#L185
We add an entry only when and later monitor it if we traced it's init execv here. Due to this an entry is never added for container even though we are able to get it's events later. We can remove this condition and add an entry initialized to 1 if we don't find it here in this condition:https://github.com/kubearmor/KubeArmor/blob/b6df53ce1a910224c3aeba117361e6b9ecf52f14/KubeArmor/BPF/system_monitor.c#L181. Something like this, 
```
    if (pid_ns_map.lookup(&pid_ns) != 0) {
        return pid_ns;
    } else {
        pid_ns_map.update(&pid_ns, &one);
    }
```
Basically we would add entry based on whether container's PID namespace has been added or not rather than whether it's init execv has been traced or not.

**Comment by nam-jaehyun** [comment]
@oneiro-naut based on your PR, just wondering if we have any side effects when we monitor the pid_ns for all processes.

**Comment by nam-jaehyun** [comment]
@oneiro-naut It may allow KubeArmor to generate logs for already-deployed containers, but the modification has a critical issue that it also generates too many non-container-related logs (e.g., containerd, kubelet, ...).

**Comment by oneiro-naut** [comment]
@nam-jaehyun Is it possible here to filter container namespaces to tell if the pid_ns added is not a container without relying on pid == 1(init) condition? Maybe if we could give specific pid_ns different value than 1 to indicate that we don't want to monitor it's events and return from it early on?

**Comment by nam-jaehyun** [comment]
@oneiro-naut we know how we can get the logs for already-deployed containers. now, it's time to investigate the side effect and develop a solution for this, which means that we need careful analysis.
